src/node_contextify.cc

```diff
@@ -852,13 +855,21 @@ class ContextifyScript : public BaseObject {
       result = script->Run();
     }
 
-    if (try_catch.HasCaught() && try_catch.HasTerminated()) {
-      env->isolate()->CancelTerminateExecution();
-      if (timed_out) {
-        env->ThrowError("Script execution timed out.");
-      } else {
-        env->ThrowError("Script execution interrupted.");
+    if (try_catch.HasCaught()) {
+      if (try_catch.HasTerminated())
+        env->isolate()->CancelTerminateExecution();
+
+      // It is possible that execution was terminated by another timeout in
+      // which this timeout is nested, so check whether one of the watchdogs
+      // from this invocation is responsible for termination.
+      if (timed_out || received_signal) {
+        if (timed_out) {
+          env->ThrowError("Script execution timed out.");
+        } else {
+          env->ThrowError("Script execution interrupted.");
+        }
       }
+
       try_catch.ReThrow();
```

Comment:
@bnoordhuis Not sure I’m completely understanding you – the `ReThrow()` is necessary because otherwise the `TryCatch` would catch the `Script execution timed out/interrupted` errors.

But you’re right, it might be a good idea to explicitly state what exactly is thrown here, I’m adding a comment.